### ISSUE TITLE
Adding Kubernetes Namespace without SCDF restart

### ISSUE THREAD

**Issue by umutcann** [spring-cloud/spring-cloud-dataflow#4390]
Hi there,

We have 2.7.0-RC1 version SCDF which is running on Kubernetes. We are storing deployable namespace(platform) in configmap like below. 

```
> spring:
>   cloud:
>     dataflow:
>       task:
>         platform:
>           kubernetes:
>             accounts:
>               namespace-a:
>                 namespace: namespace-a
>                 entryPointStyle: shell
>                 imagePullPolicy: Always
>                 limits:
>                   memory: 2048Mi
>               namespace-b
>                 namespace: namespace-a
>                 entryPointStyle: shell
>                 imagePullPolicy: Always
>                 limits:
>                   memory: 1024Mi
>               ......
```

In our community, we are generating new namespace for batch apps by our need. When every new namespace has generated, we are appending new namespace manually and restarting Spring Cloud Data Flow for our 4 environment(dev,test,qa,prod).

So we want to add namespace(platform) directly on dashboard or API. Is there any feature or API to add namespace to Spring Cloud Data Flow without restart?

Thanks.

**Comment by sabbyanandan** [comment]
Hi, @umutcann. Today, the platform accounts (in SCDF/Skipper) are represented as declarative configurations at the time of installation, and they are _not_ persisted. We persist just the stream/task manifest and what platform they were deployed against. 

Feel free to explore how to design this persistence component, and start a dialog with the PR.

**Comment by deepak5127** [comment]
Hi @sabbyanandan  . Shall i give a try on this one ?

**Comment by deepak5127** [comment]
@sabbyanandan Thinking of creating an UI design flow like we have for App registration where users can add/edit Kubernetes / CF configuration and persist to the datastore. Please let me know if this is approach is fine

**Comment by sabbyanandan** [comment]
I could something like that working from a UX perspective. When you get a chance, it'd be great to see the flow in some diagram format. A hand-drawn paper diagram (or anything simpler than this) would be sufficient.

**Comment by deepak5127** [comment]
@sabbyanandan  I am attaching a very high level flow diagram. Please let me know if you want me to elaborate on this.
![IMG_0543](https://user-images.githubusercontent.com/360576/113638166-154c1580-962b-11eb-9081-01a7673f8cdb.jpg)
![IMG_0542](https://user-images.githubusercontent.com/360576/113638170-1715d900-962b-11eb-89ed-3f7c6fd9d635.jpg)

**Comment by sabbyanandan** [comment]
I like what I see, esp. the UI workflow as to how users can discover the persisted platforms and then be able to override and/or adjust the platform-specific global deployer properties. To be clear, though, "namespace" happens to be one of the many deployer properties that one can override/set globally for Kubernetes deployment.

You will find a few [use-case scenarios in the Mircosite](https://dataflow.spring.io/docs/recipes/multi-platform-deployment/multiple-platform-accounts/) where such global settings would be useful.

So, that said, we should be able to open up all applicable deployer properties (see: [cf](https://docs.spring.io/spring-cloud-dataflow/docs/current/reference/htmlsingle/#configuration-cloudfoundry-deployer) / [k8s](https://docs.spring.io/spring-cloud-dataflow/docs/current/reference/htmlsingle/#configuration-kubernetes-deployer)) ) and not just the "namespace".

**Comment by deepak5127** [comment]
Sorry I misinterpreted the flow.

So it should be like this.

The users would be able to add / edit / remove the platforms. when they add/edit the platform accounts, there will be an option to dynamically check if there is any new namespace available ( Kind of dropdown we can provide in the form ). Similarly we can see if any other properties value can be populated dynamically, like secrets from K8s  ?

**Comment by sabbyanandan** [comment]
Yep, that makes sense to me. We should be able to give them visibility into all global properties that someone can discover/manage with the help of platform accounts. We already show those deployer properties when deploying a stream (or when launching tasks in 2.8.x/master codebase, not shipped yet); maybe we could repurpose that endpoint to populate all the applicable properties here also.

Nicely done, @deepak5127! 

@ilayaperumalg / @jvalkeal, what do you think of this approach? Do you have any feedback for @deepak5127?

**Comment by ilayaperumalg** [comment]
@deepak5127 Thanks for your interest and proposal on this!

We need to have the similar design on the Skipper server as well. Currently, the platform accounts for tasks are maintained in SCDF while the platform accounts for streams are driven by Skipper server. Hence, ability to dynamically configure the platform accounts need to be supported at both the Skipper and SCDF servers. 

Also, when it comes to UX design, along with the task platform accounts configuration SCDF server UI could allow the update of stream platform accounts that eventually get updated at Skipper server.

**Comment by deepak5127** [comment]
@ilayaperumalg  @sabbyanandan  Thanks for the feedback. Will incorporate the skipper server in the flow. Should i start designing the DB schema ?

**Comment by ilayaperumalg** [comment]
Sure, please go ahead! Thank you

**Comment by deepak5127** [comment]
@sabbyanandan  @ilayaperumalg  Attaching the DB table design. Please review and provide feedback.

![scdf](https://user-images.githubusercontent.com/360576/114084953-97c01980-9865-11eb-874d-1be3789d3a9d.png)

**Comment by ilayaperumalg** [comment]
Hi @deepak5127 , 

Thanks for the proposal!

I'd like to know the reason for having `PLATFORM` table. We have platform type which can differentiate the type of platform for the given platform account and the configs can be associated with each of the `PLATFORM_ACCOUNT` entry. 

We can also rename the table name `ACCOUNT_CONFIG_KV` to `PLATFORM_ACCOUNT_CONFIGS` or something similar?

Also, with this approach we need to convert the Boot configuration properties (Platform properties that correspond to the underlying deployer properties - K8s, CF, local) and have them converted into key/value to persist in the `ACCOUNT_CONFIG_KV` table. Do you have specific idea in mind to do this? 

Since all the account properties at SCDF correspond to task platform accounts, we would need to consider having a prefix TASK (something like `TASK_PLATFORM_ACCOUNT`) in all these tables.

**Comment by ilayaperumalg** [comment]
Adding @jvalkeal to add his comments on the schema changes.

**Comment by deepak5127** [comment]
@ilayaperumalg  The PLATFORM table was added to have more integrity at the DB level. We can eliminate PLATFORM table if its not necessary. 

On the table name convention, Yes we can rename the tables accordingly. 

On storing the properties, there are two options.

1 . We can store it has simple BLOB or CLOB. This implementation is less complex, as we dont have to worry about schema changes to the JSON document ( properties ). We might have to update the entire JSON doc whenever there is a change in properties which I think is ok. 
2. We can store each property in a row. This is complex compared to first approach, as we need to design how to store property which can accept multiple values.  

I have designed the DB schema based on 2nd approach, but I think we should just try to keep simple and go with approach one, as our use case needs just a persistent store of properties value. Approach 2 might be helpful when users want to update the properties value directly in the DB. If we decide to go with approach 2, then we need only two tables. TASK_PLATFORM_ACCOUNT and TASK_PLATFORM_ACCOUNT_CONFIGS.

Please share your thoughts.

**Comment by ilayaperumalg** [comment]
Thanks @deepak5127 

The above discussion reminds me of one another important point we discussed in the past on mutating the platform account properties.

Originally we would like to have the platform account properties to be **immutable**.  
The reason for this is to preserve the integrity of the existing streams/tasks that are used by the platform accounts which might get changed upon the server restart or with the current proposal in the PR. In this case, the platform accounts require maintaining the change history of the platform accounts for the existing stream/task deployments. 

I guess we need to discuss this first before doing a deep dive on the schema itself. 

I still propose having the platform accounts immutable and if there is a need user could add a new platform account with the required properties as it wouldn't affect the existing stream/task deployments using the previous platform account settings.

cc @cppwfs

**Comment by deepak5127** [comment]
Sure @ilayaperumalg  . I will wait for the direction.

**Comment by umutcann** [comment]
@ilayaperumalg is there any update on that enhancement request? Have you put it to SCDF roadmap?
